### PR TITLE
WL-3732 : New groupstore

### PR DESCRIPTION
**Pull request by nicholaswilson100**

### CODE REVIEW

**Comment by nicholaswilson100** [comment]
Linked to https://github.com/ox-it/wl-providers/pull/5

These might need to be merged at the same time to avoid Jenkins carnage

**Comment by nicholaswilson100** [comment]
I've raised the search limit to 500 (ten times the current 50) because we are getting back ten times the results from a user hitting the Search button now or from autocomplete running.  From testing this seems to work okay for most cases.  After the results are got back from ldap, I'm filtering them in the Java to show only the ones we're interested in (e.g. no suspended groups, no Return contract groups,etc).

I wanted to do all the filtering in the LDAP search filter but LDAP doesn't let you do wildcard or regex :(

**Comment by nicholaswilson100** [comment]
The change to search.jsp is so we see more of the tree browser on the screen

**Comment by nicholaswilson100** [comment]
We are adding a suspended group for each non-suspended course group that gets added for a site. 
So this needs a migration script along side it to backfill sites that don't have a suspended group as well.

**Comment by nicholaswilson100** [comment]
A start to the Migration script can be found here: https://jira.oucs.ox.ac.uk/jira/browse/WL-3894

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 1]
```diff
@@ -214,6 +275,9 @@ private boolean isValidTerm(String term) {
 			LDAPSearchConstraints constraints = new LDAPSearchConstraints(connection.getConstraints());
 			constraints.setMaxResults(SEARCH_LIMIT);
 			connection.setConstraints(constraints);
+            searchPattern = "(&{0}(member=*)(objectClass=groupstoreGroup))";
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: Indentation looks wrong.

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 2]
```diff
@@ -35,6 +37,14 @@
 	JLDAPDirectoryProvider jldapDirectoryProvider;
 	
 	UserDirectoryService userDirectoryService;
+
+	private MemoryService m_memoryService;
+	public void setMemoryService(MemoryService memoryService) {
+		this.m_memoryService = memoryService;
+	}
+	public MemoryService getM_memoryService() {
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: Getter looks wrongly named? Do you need the getter at all?

File: api/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManager.java [hunk 3]
```diff
@@ -47,4 +47,11 @@
 	 */
 	public String findExternalGroupId(String mappedGroupId);
 
+	/**
+	 * Gets and caches the filter for searching LDAP for course owners
+	 * @param cacheName The name of the cache
```
Comment (buckett) [inline api/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManager.java]: What's the reason for wanting to allow the caller to cache it under a different key.

**Comment by buckett** [comment]
I don't see a migration script in: https://jira.oucs.ox.ac.uk/jira/browse/WL-3894

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 4]
```diff
@@ -106,7 +135,39 @@ public String findExternalGroupId(String mappedGroupId) {
 		MappedGroup mappedGroup = mappedGroupDao.findById(mappedGroupId);
 		return (mappedGroup == null)?null:mappedGroup.getExternalGroup();
 	}
-	
+
+	public String getCourseOwnerFilter(String cacheName) throws ExternalGroupException {
+		String courseOwnersFilter;
+		if (courseOwnersCache.containsKey(cacheName)) {
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: There is no point doing the containsKey check as there's no guarantee it won't expire by the time you get it out.

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 5]
```diff
@@ -106,7 +135,39 @@ public String findExternalGroupId(String mappedGroupId) {
...
+			}
+			oxfordCourseOwners = "(|" + oxfordCourseOwners + ")";
+		} catch (LDAPException lde) {
+			log.warn(lde);
+			throw new ExternalGroupException(Type.UNKNOWN);
+		} finally {
+			returnConnection(conn);
+		}
+
+		courseOwnersCache.put(cacheName, oxfordCourseOwners);
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: If we don't get any results back (eg no permission to read that part of the tree), do we want to cache that for an hour?

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 6]
```diff
@@ -353,25 +426,29 @@ public void setDisplayNames(Map<String, String> displayNames) {
...
+						parseValues = formatter.parse(member);
+						if(parseValues.length == 1) {
+							String eid = (String) parseValues[0];
+							users.add(eid);
+						} else {
+							log.warn("Failed to parse member of group: "+ member);
+						}
+					}
+					catch (ParseException e){
+						log.warn("It's a grouper group" + e.getMessage());
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: If this is expected we don't want to log a warning.

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java (hunk 2 above)
Comment (nicholaswilson100) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: I'll change the name so it's less odd-looking. 
The getter is needed / called (although I would have thought Spring only needed the setter).

File: api/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManager.java (hunk 3 above)
Comment (nicholaswilson100) [inline api/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManager.java]: Good point

**Comment by nicholaswilson100** [comment]
I still need to do the migration script

File: impl/src/main/java/uk/ac/ox/oucs/vle/AttributePathHandler.java [hunk 7]
```diff
@@ -26,21 +21,34 @@
...
 	private String attribute;
-	
+
+	private String attribute2;
+
 	private String displayAttribute;
 	
 	private String searchAttribute;
 
+	public static final String OXUNI = "oxuni";
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/AttributePathHandler.java]: Generally the Handlers were all free of Oxford config and the setup/config was passed to them in the constructor. Now the harded coded strings are spread out around the classes more.

Comment (nicholaswilson100) [inline impl/src/main/java/uk/ac/ox/oucs/vle/AttributePathHandler.java]: Ok. I'll have a look at this.

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 8]
```diff
@@ -214,6 +277,9 @@ private boolean isValidTerm(String term) {
 			LDAPSearchConstraints constraints = new LDAPSearchConstraints(connection.getConstraints());
 			constraints.setMaxResults(SEARCH_LIMIT);
 			connection.setConstraints(constraints);
+			searchPattern = "(&{0}(member=*)(objectClass=groupstoreGroup))";
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: Why are you updating the instance variable here? Isn't it already set to this?

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java [hunk 9]
```diff
@@ -345,7 +420,7 @@ public void setDisplayNames(Map<String, String> displayNames) {
 		LDAPConnection connection = null;
 		try {
 			connection = getConnection();
-			LDAPSearchResults results = connection.search(externalId, LDAPConnection.SCOPE_BASE, null, new String[]{memberAttribute}, false);
+			LDAPSearchResults results = connection.search(externalId, LDAPConnection.SCOPE_BASE, "member=*", new String[]{memberAttribute}, false);
```
Comment (buckett) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: Is this just to shortcut the loop?

**Comment by buckett** [comment]
Keeping all the Oxford specific stuff in one place makes it easier if we decide to contribute this back as we have to generic modules that all people to browse parts of an ldap tree.

**Comment by buckett** [comment]
Did you have any performance numbers about how long the search for owners takes and how many are returned?

File: impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java (hunk 8 above)
Comment (nicholaswilson100) [inline impl/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManagerImpl.java]: Aah. Should have removed it after using it for testing.

File: impl/src/main/java/uk/ac/ox/oucs/vle/AttributePathHandler.java (hunk 7 above)
Comment (nicholaswilson100) [inline impl/src/main/java/uk/ac/ox/oucs/vle/AttributePathHandler.java]: I was thinking that ExternalGroupManagerImpl might be a possible place to add the constants.  It already has some oxford config in it.

**Comment by nicholaswilson100** [comment]
Yes I'll get some performance numbers for the owners

**Comment by nicholaswilson100** [comment]
I've attached a script to https://jira.oucs.ox.ac.uk/jira/browse/WL-3732 for migrating old DNs in MAPPED_GROUP.EXTERNALGROUP to groupstore DNs.

**Comment by nicholaswilson100** [comment]
I tested it by creating a bunch of sites and adding  different course groups to those sites, then ran the migration, checked out my branch and looked to see if the groups were still there (and still had roughly the same members/member numbers).

**Comment by nicholaswilson100** [comment]
There are 4 static identical tables because in mysql 'You cannot refer to a TEMPORARY table more than once in the same query. ' 

http://dev.mysql.com/doc/refman/5.0/en/temporary-table-problems.html

There are other ways to have done this but as the table was so small it seemed fine.

**Comment by nicholaswilson100** [comment]
The owners search which is this:

ldapsearch -x -Z -h ldap.oak.ox.ac.uk -b ou=course,ou=groupstore,dc=oak,dc=ox,dc=ac,dc=uk -s sub oxfordCourseOwner=* oxfordCourseOwner

returns 2094 owners.

Timings-wise, I get this which seems okay (for a one time cache):

$ time ldapsearch -x -Z -h ldap.oak.ox.ac.uk -b ou=course,ou=groupstore,dc=oak,dc=ox,dc=ac,dc=uk -s sub oxfordCourseOwner=* oxfordCourseOwner > /dev/null

real	0m1.450s
user	0m0.135s
sys	0m0.050s

**Comment by nicholaswilson100** [comment]
Linked to https://github.com/ox-it/wl-kernel/pull/66

**Comment by nicholaswilson100** [comment]
Linked to https://github.com/ox-it/wl-site-manage/pull/39

File: api/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManager.java [hunk 10]
```diff
@@ -47,4 +47,10 @@
 	 */
 	public String findExternalGroupId(String mappedGroupId);
 
+	/**
```
Comment (buckett) [inline api/src/main/java/uk/ac/ox/oucs/vle/ExternalGroupManager.java]: I don't think this needs to be in the API or should be.
Whey would an external module need to call this?

**Comment by nicholaswilson100** [comment]
Removed getCourseFilter() from API and done a rebase.

**Pull request merged**